OpenSim/Simulation/Model/CMCActuatorSubsystem.cpp

```diff
@@ -148,21 +148,28 @@ void CMCActuatorSubsystemRep::setSpeedTrajectories(FunctionSet *aSet) {
     // controls
     const CoordinateSet& coords = _model->getCoordinateSet();
     for (int i = 0; i < nq; ++i) {
-        // the last argument to setValue is a bool to enforce kinematic constraints
-        // or not. It is being set to true when we set the last coordinate value.
-        coords[i].setValue(mutableCompState, _qWork[i] + _qCorrections[i], i==(nq-1));
+        // the last argument to setValue, a bool to enforce constraints,
+        // is false since values come from a _qSet of splined desired
+        // kinematics formed from formCompleteStorages, which enforces
+        // model constraints.
+        coords[i].setValue(mutableCompState, _qWork[i] + _qCorrections[i], false);
         coords[i].setSpeedValue(mutableCompState, _uWork[i] + _uCorrections[i]);
     }
-
-     /* copy  muscle states computed from the actuator system to the muscle states
-        for the complete system  then compute forces*/
+    // project() to satisfy constraints perturbed by _q/_uCorrections
+    _model->getMultibodySystem().projectQ(mutableCompState,
+        getModel()->get_assembly_accuracy() / 10);
+    _model->getMultibodySystem().projectU(mutableCompState,
+        getModel()->get_assembly_accuracy() / 10);
+
+    /* copy  muscle states computed from the actuator system to the muscle states
+       for the complete system  then compute forces*/
```

Comment:
Is this different from assemble? If so, is it a facility that needs to be exposed at a higher level? Ideally developers of similar tools don't need to go to this low level multibody system calls if the're necessary.

Reply:
Projection is not an assembly. It does not need to be exposed at a higher level. This is similar to what the integrator does to satisfy kinematic constraints as time advances. Here, the kinematic state is being slightly altered by` _qCorrections` and `_uCorrections` and the projection is ensuring that the update state still satisfies constraints.

The `CMCActuatorSubsystem` does not integrate the multibody dynamics, just muscle dynamics, and by not doing the projection it can potentially use kinematic states (which it holds constant) that are not valid in estimating muscle controls. When the actual system is integrated, the constraints are satisfied during integration so we do not see constraint, but it is potentially using an internal state that is not accurate. This likely explains why the solutions are: 1) different and 2) slower (but not as slow as when we were doing an assembly).